controllers/sandbox_controller.go

```diff
@@ -266,13 +291,36 @@ func (r *SandboxReconciler) reconcileService(ctx context.Context, sandbox *sandb
 func (r *SandboxReconciler) reconcilePod(ctx context.Context, sandbox *sandboxv1alpha1.Sandbox, nameHash string) (*corev1.Pod, error) {
 	log := log.FromContext(ctx)
 	pod := &corev1.Pod{}
-	if err := r.Get(ctx, types.NamespacedName{Name: sandbox.Name, Namespace: sandbox.Namespace}, pod); err != nil {
+	err := r.Get(ctx, types.NamespacedName{Name: sandbox.Name, Namespace: sandbox.Namespace}, pod)
+	if err != nil {
 		if !k8serrors.IsNotFound(err) {
 			log.Error(err, "Failed to get Pod")
 			return nil, fmt.Errorf("Pod Get Failed: %w", err)
 		}
-	} else {
+		pod = nil
+	}
+
+	// if replicas is 0, delete the pod if it exists
+	if *sandbox.Spec.Replicas == 0 {
+		if pod != nil {
+			if pod.ObjectMeta.DeletionTimestamp.IsZero() {
+				log.Info("Deleting Pod because .Spec.Replicas is 0", "Pod.Namespace", pod.Namespace, "Pod.Name", pod.Name)
+				if err := r.Delete(ctx, pod); err != nil {
+					return nil, fmt.Errorf("failed to delete pod: %w", err)
+				}
+			} else {
+				log.Info("Pod is already being deleted", "Pod.Namespace", pod.Namespace, "Pod.Name", pod.Name)
+			}
+		}
+		return nil, nil
+	}
+
+	if pod != nil {
 		log.Info("Found Pod", "Pod.Namespace", pod.Namespace, "Pod.Name", pod.Name)
+		if !pod.ObjectMeta.DeletionTimestamp.IsZero() {
+			// Sometimes when we delete and recreate a sandbox quickly, the old pod might still be terminating
+			log.Info("Pod is being deleted, will wait for it to be gone before recreating", "Pod.Namespace", pod.Namespace, "Pod.Name", pod.Name)
```

Comment:
This _looks_ wrong, like if we're doing something special here shouldn't we be doing more than logging?  Should we at least be scheduling a requeue?  Or is the idea that we are already watching the pod?

Reply:
So maybe we don't need to special-case log it here. We assume the pod.phase should reflect it. 
The idea is we are watching the pod(s). And we return the pod so that the sandbox ready condition captures the pod phase as is. (for all cases other than ready).

Reply:
removing the special case if here since it is not needed and the code is robust as it without it.